CLIENT_SPECIFICATION.md

```diff
@@ -0,0 +1,194 @@
+# tldr-pages client specification
+
+ - **Current Specification Version:** 2.0
+
+This document contains the official specification for tldr-pages clients. It is _not_ a specification of the format of the pages themselves - only a specification of how a user should be able to interface with an official client. 
+
+The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+
+## Changelog
+ - v1.0 (18th December 2018)
+	 - Initial draft
+ - v1.1 (18th December 2018)
+	 - Refined bullet points according to comments
+	 - Ensure the spec specifies the _interface_, not the workings
+ - v2.0 (12th January 2019)
+	 - Expanded notes into full sentences
+
+
+## Terminology
+In order to aid the understanding of this specification document, a number of terms will be defined in this section.
+
+### Page
+tldr-pages consists of multiple _pages_ - each of which describes a specific command.
+
+### Platform
+Pages are grouped by platform. A platform is an operating system. For example, `windows`, `linux`, `osx`. The special platform `common` contains pages for commands that work identically across all platforms.
+
+
+## Command-Line Interface
+This section describes the standardised command-line interface (CLI) for clients implementing one. Clients that do not provide a CLI can ignore this section.
+
+### Arguments
+A number of arguments MUST be supported (unless otherwise specified if a CLI is implemented:
+
+Argument		        | Meaning
+------------------------|--------------------
+`--update`				| Updates the offline cache of pages. MUST NOT be implemented if cache is not supported.
+`--version`, `-v`		| Shows the current version of the client, and the version of this specification that it implements.
+`--list-all`, `-a`		| Lists all the pages in the current platform to the standard output. One page name per line. Additional decoration MAY be printed to the standard error.
```

Comment:
how about `--list`, `-l`? This way it is similar to `ls -l` (one per line), and shorter to type.

Reply:
Hrm, you raise a good point. What do you think, @agnivade / @pxgamer / @waldyrious?

Reply:
I think this was because the node client uses `--list-all` :thinking_face:

Reply:
I think `--list-all` is rather confusing, actually. It implies that all pages are listed, but in reality it's only the pages in the current platform & common that are listed.

Perhaps we can disambiguate this somehow?

Reply:
Personally I don't see any value in listing every possible page, crossplatform (which I would assume `--list-all` does as @jedahan pointed out) because that means we should print out paths to disambiguate pages covering same tool on different platforms. That essentially puts us in business of re-implementing `ls -R`. If user really wants that they should be able to get cache location somehow and do that on they own.
So my vote goes to `-l, --list` to get list of pages for assumed/specified platform.

OTOH discussion about list formatting is (sorry but I have to say it) aiming at completely wrong target. It simply **does NOT matter** at all how it will look and there is good background reason for that. Take our beloved `ls` for example. Default output of `ls` is columned or is it?! Pipe that through `cat/less` and you'll get _one-per-line_ list. I'm too lazy to dig out proper POSIX/TAOUP/whatever spec/recommendation so rule of thumb instead:

Output of _first class citizen_ *NIX CLI tool:
1. **MAY** be aesthetically pleasing with all kind of wonders you decide to throw in
**if `STDOUT` is NOT `TTY`**
2. **MUST** be formatted as _one-per-line_ list (to enable easy manipulation using *NIX tools)
**if `STDOUT` is a `TTY`**

Can we just replace _blah blah CLI tool_ with `tldr` client and put it inside spec? :wink:

Reply:
I think listing all pages are useful to know what all commands are there (all platforms or current platforms comes after). This is a tool aimed for beginners and we should not expect them to go to the cache directory and manually check the folder.

Now regarding listing pages from all platforms or a single platform is up for debate. I think it's reasonable to want to see _all_ pages from all platforms, otherwise people will need to type all the platforms along with `--list` to get all the pages. Disambiguation will be needed, yes.

In any case, `--list-all` should mean display all pages from all platforms. `--list` should be for current platform and can be combined with `--platform` flag.

Reply:
If you envision future where `tldr --list --platform macos` is a thing, can we just do `tldr --list --platform all`?

Reply:
I would be okay with `all`, but do you think there are other places where we can use `all` or just for this ? If it's just for this, then why not change `--list` to mean `--list-all` ? Better not to attach special meaning to any word, if we can keep things simple.

Reply:
Well, that's just my opinion but I give more importance to listing current platform pages than to dumping everything and options should then follow that assumption. Meaning `-l,--list` does that unless you add _pro_ type of option that will turn it into dump everything. Most useful option should be easiest to type.

Reply:
Got it. I am kinda opposite to that notion. I think list (or whatever) should dump all platforms by default until I want just a specific platform.

So I propose - `-l,--list` shows all platforms by default, unless you add a `--platform` flag, upon which it just shows pages for that platform.

That is just my opinion. We will decide whatever the community wants.

Reply:
Ah that makes sense. Thanks.

Reply:
if `tldr --list` with no platform defaults to all, that is the inverse behaviour of `tldr <command>` with no platform defaulting to the detected platform.

Reply:
Great points.

I think @jedahan has a point about consistency. What about `tldr --list` defaults to the current platform, `tldr --list platform` lists the pages on a specific platform, and `tldr --list all` lists all pages inn all platforms? While I like the idea of using a `*`, that could be interpreted by the shell to mean "all files in the current directory" and get substituted in, so I'd like to avoid that if possible.

I'm unsure how the `common` platform fits into all of this though.

Great point @vladimyr on if the output is a TTY. I think I'll add that.

Reply:
@sbrl Depends on language/tooling you use but this list covers pretty much all popular options: https://rosettacode.org/wiki/Check_output_device_is_a_terminal In, a nutshell there is either `isTTY(fd)` utility function provided or a property on `stdout` itself.

Reply:
I still think if people want to do something that technical, they could either just turn to the cache directory and get anything they need there, or, they should use the `-t`/`--tty` flag or `-n`/`--noformat` or some such. Because if people pipe their output into `less`, they probably don't want to automatically get no columns or colors. If you use a flag, the user can decide, if you don't, some users are going to request it as a feature.

Reply:
@pepa65

> If you use a flag, the user can decide, if you don't, some users are going to request it as a feature.

And that's like only thing I agree with. :disappointed:

> I still think if people want to do something that technical,

Calling list manipulation with standard cli tools technical is bit far fetched if you ask me; after all isn't one of selling points of `tldr` pages is to get people to that level more easily? :upside_down_face:
What so technical in grepping large lists? :thinking:

> or, they should use the `-t`/`--tty`

I'm very curious to see example of tool implementing given option. I never seen something like that and I highly doubt I ever will; having in mind it is not considered standard:
- http://www.catb.org/esr/writings/taoup/html/ch10s05.html
- https://www.gnu.org/prep/standards/html_node/Option-Table.html#Option-Table
    _(`tty` used by `gdb` is not switch but an option with an argument: https://manpages.ubuntu.com/manpages/precise/en/man1/gdb.1.html)_

>Because if people pipe their output into less, they probably don't want to automatically get no columns or colors.

Sorry but that's just your opinion while common practice teaches us exact opposite. I may or may not agree with you but you are actually arguing with **facts**:
- https://superuser.com/questions/36022/getting-colored-results-when-using-a-pipe-from-grep-to-less
- https://unix.stackexchange.com/questions/19317/can-less-retain-colored-output

So, yeah you are right, people like ansi colors but _like_ and _expect_ by default are two different things. By default colors are off. Which brings me to the next thing you said:

> or -n/--noformat or some such.

Again, you are going against common practice established by *NIX cli tools. Formatting is off by default if output is tty. If you really want to retain it you need to explicitly request tool generating output to do so. And it makes perfect sense why you need extra step for colors/formatting to be outputted/preserved: **not every tool you may use to process output understands and/or is able to strip out ansi escapes!** If user is positive that his tool of choice (e.g. `less`) is capable of doing so, ok, you have choice to do `--color=always`, `--color`, `--pretty`, you name it...

In conclusion - to actually address your concern we should probably support `--pretty` option that would force colors and general formatting being preserved when output is not `tty`. I'm ok with that and FWIW you have my :+1: there but in form of recommendation, such thing shouldn't be a **MUST** but **MAY** instead:
- `stdout is NOT TTY` - Generate formatted output _(mandatory)_
- `stdout is TTY and --pretty` - Preserve formatting _(optional)_
- `stdout is TTY` - Generate `\n` separated list _(mandatory)_

That being said, please try to align your proposals with established CLI tools design practices and user expectations. TBH I understand your motivation and would sometimes give everything for a chance to _rewrite_ history of our discipline but that's simply not possible. `tldr` clients are not first tools of such class and less exotic they are more successful and useful they'll become!

Reply:
@pepa65
>Don't you feel you are being unnecessarily redundant with `tldr --list --platform macos`? I just do `tldr --list macos` and it works. There is no need there for a `--platform` marker.

Yes and no :thinking:

Sure it is verbose, I personally don't like it too but...
OTOH lets compare this with what `man` does (I like to consider `tldr` client as _`man` for humans_ same way as [`httpie`](https://httpie.org)  is _`curl`for humans_; correct me if my understanding is wrong :bowtie:).

In order to get all `man` pages you need to run:
```
$ man -k .
```
where `.` is regex meaning everything. If you want to narrow that down using sections:
```
$ man -k . -s1
```
lists only manpages inside 1st section.

If you now swap `-k .` with `-l` and `-s1` with `-p <platform>` you'll get original form I proposed:
```
$ tldr -l -p macos
$ # or if you like it more
$ tldr --list --platform macos
```

You could OFC argue that manpage listing I just described has mandatory search param while tldr listing does not thus putting us in better position by allowing what you suggested. Yeah, that's true but might be also an indication that we are doing something wrong :warning:
Maybe right way to do the listing is to copy good ol' man? :upside_down_face:

@jedahan
Could not support you more in that consistency effort :100:

@sbrl
I wanted to do some wildcard like thingie and ended up doing lousy job :man_facepalming:

I know it isn't constructive at all but this whole thing with what should _list (almost) everything_ print is painfully convoluted and more we evolve it less value I see in it. Can someone ELI5 in what particular case user wants to print out every or almost every possible command out apart from pure curiosity and explorer's spirit?
Aren't the _real life_ usecases:
(step zero: you know what you want to do)
1. You **know** which tool can help you but you don't know how to use it so you do:
```
$ tldr <tool>
```
2. You **do NOT know** which tool you should use so you do:
```
$ tldr --search <query>
```
**Which isn't even mentioned in the spec itself!** :warning:

In both cases you might be searching answers for a platform you are currently not working on so you'll add `--platform <target_platform>` too.

I mean this is exactly what `man` supports also:
```
$ man <tool>
$ man -k <query> # analogous to `--search`
```
while listing all pages is available by means of _empty_ query but is certainly isn't something that is endorsed by having dedicated option like `--list-(almost/maybe/platform/common)-all` :upside_down_face:

Reply:
Alright then, let's go with default being current platform and having a reserved platform `all`.

Reply:
Great discussion! I'll update the spec to reflect this.

Met me know if I miss anything :P